apps/docs/docs/nest.md

```diff
@@ -28,6 +29,42 @@ The `@Api` decorator takes the route, defines the path and method for the contro
 
 It also injects "appRoute" into the req object, allowing the `@ApiDecorator` decorator automatically parse and check the query and body parameters.
 
+
+## Response Return Type Safety
+
+You have two options to ensure HTTP type safety on your Nest Controllers: 
+
+- `ControllerShape` as shown above: 
+  - Your controller can implement the `ControllerShape` derived from `typeof s.controllerShape`
+  - This ensures your controller methods also align with the base interface shapes
+- `ResponseShapes`:
+  - Your controller can utilize `typeof s.responseShapes` in the return type. For example:
+    ```typescript
+    const s = initNestServer(apiBlog);
+    type ControllerShape = typeof s.controllerShape;
+    type RouteShape = typeof s.routeShapes;
+    type ResponseShapes = typeof s.responseShapes; // <- http Responses defined in contract
+
+    @Controller()
+    export class PostController {
+      constructor(private readonly postService: PostService) {}
+
+      @Api(s.route.getPost)
+      async getPost(@ApiDecorator() { params: { id } }: RouteShape['getPost']):
+      Promise<ResponseShapes['getPost']> { // <- return type defined here
+        const post = await this.postService.getPost(id);
+
+        if (!post) {
+          return { status: 404, body: null };
+        }
+
+        return { status: 200, body: post };
+      }
+    }
+    ```
+  - If your controller needs to implement a difference class, or needs extra methods defined outside of a contract, this is option gives that flexibility
```

Comment:
I believe this is not entirely accurate. You should be able able to define extra methods other than than ones in the contract and typescript would not complain. You can also implement multiple interfaces or extend another class and also still implement the interface, i.e., `class Controller extends BaseController implements ControllerShape, AnotherInterface`

The only case I could see this useful if you are not fully implementing the entire contract in a single controller. But even then, you can use Pick/Omit to achieve the same goal, so something like `class Controller implements Pick<ControllerShape, 'getPost'>`

I'm not sure if it's worth introducing this change to the Nest plugin as it could be a bit confusing as to what approach should be taken by someone reading the documentation. Also the explicit typing in the method declaration is quite unwieldy

`async getPost(@ApiDecorator() { params: { id } }: RouteShape['getPost']): Promise<ResponseShapes['getPost']>`

Reply:
```
You can also implement multiple interfaces or extend another class and also still implement the interface, i.e., class Controller extends BaseController implements ControllerShape, AnotherInterface
```
This is true, I can update this verbiage. However, overall, observing the specific TS error you're getting with a class-based approach is exponentially difficult. For example, take the `posts.controller` file:

<img width="860" alt="Screenshot 2022-12-29 at 8 10 02 AM" src="https://user-images.githubusercontent.com/55844504/209955694-c4bafbe5-7551-4eb0-95bb-03429b97c914.png">

To see the exact TS error you're receiving, you have to do quite a bit of scrolling to see what's wrong.  Over a significantly sized controller, this approach, in my opinion, isn't scalable. Extending a class and finding graining with `class Controller implements Pick<ControllerShape, 'getPost'` is also quite a bit of extra work and not as intuitive in my opinion. It also promotes extra inheritance patterns, which, as we all know, can get quite messy (since TS allows for 3rd level extensions/inheritance). 

```
I'm not sure if it's worth introducing this change to the Nest plugin as it could be a bit confusing as to what approach should be taken by someone reading the documentation. Also the explicit typing in the method declaration is quite unwieldy
```
Again, this is all preference/opinion, which is why it's more of an addition than a change. If you'd both like, I can mark the title in the docs`Optional` for this approach. I know some users would definitely appreciate this feature, including my company.

Reply:
Aha! Yeah, that totally makes sense now, I definitely agree! Now that I'm checking, it's pretty much the same issue with all of the server libraries.